### ISSUE TITLE
Failed to bind to address https://127.0.0.1:5001: address already in use

### ISSUE THREAD

**Issue by jenyckee** [dotnet/core#2001]
Trying to run a freshly created project I get the following. Seems like a race condition because port 5001 is not in use on my system.

```
[41m[1m[37mcrit[39m[22m[49m: Microsoft.AspNetCore.Server.Kestrel[0]
      Unable to start Kestrel.
System.IO.IOException: Failed to bind to address https://127.0.0.1:5001: address already in use. ---> Microsoft.AspNetCore.Connections.AddressInUseException: Address already in use ---> System.Net.Sockets.SocketException: Address already in use
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransport.BindAsync()
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransport.BindAsync()
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.<>c__DisplayClass22_0`1.<<StartAsync>g__OnBind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.LocalhostListenOptions.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.AddressesStrategy.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(IServerAddressesFeature addresses, KestrelServerOptions serverOptions, ILogger logger, Func`2 createBinding)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
Microsoft.AspNetCore.Server.Kestrel:Critical: Unable to start Kestrel.

System.IO.IOException: Failed to bind to address https://127.0.0.1:5001: address already in use. ---> Microsoft.AspNetCore.Connections.AddressInUseException: Address already in use ---> System.Net.Sockets.SocketException: Address already in use
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransport.BindAsync()
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransport.BindAsync()
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.<>c__DisplayClass22_0`1.<<StartAsync>g__OnBind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.LocalhostListenOptions.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.AddressesStrategy.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(IServerAddressesFeature addresses, KestrelServerOptions serverOptions, ILogger logger, Func`2 createBinding)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
The thread 1215731 has exited with code 0 (0x0).
The thread 1215727 has exited with code 0 (0x0).
Exception thrown: 'System.IO.IOException' in System.Private.CoreLib.dll

```

**Comment by kamleshrao** [comment]
There is already this issue logged.  Did you check the solution of checking Port Usage (by some other application) and killing it? 

https://github.com/dotnet/core/issues/948 

Are you on HTTPS on localhost?

**Comment by karelz** [comment]
Duplicate of #948

**Comment by jenyckee** [comment]
Okay, but what is the solution?

**Comment by karelz** [comment]
@jenyckee did you check the primary issue?

**Comment by jenyckee** [comment]
Yes

**Comment by anhmiuhv** [comment]
This also happens to me

**Comment by reevelau** [comment]
I had the same issue on a dotnet generate web project recently. But I can fix the issue by editing `Properties\launchSettings.json` to modify the `applicationUrl` in your app config.

From this
```
"applicationUrl": "https://localhost:5001;http://localhost:5000",
```
to this
```
"applicationUrl": "http://localhost:5000",
``` 

I suspect it could be an issue of the dotnet boilerplate project "ASP.NET Core Web API".

**Comment by jenyckee** [comment]
@reevelau This doesn't help for

**Comment by karelz** [comment]
> @jenyckee  did you check the primary issue?
> ...
> Yes, when I "Start Without Debugging" the application even crashes

See https://github.com/dotnet/core/issues/948#issuecomment-336284252 ... unless the right team knows about the problem and understands the root-cause, they can't provide solutions.

**Comment by jenyckee** [comment]
@karelz That's very unfortunate, what kind of information do you need in order to understand the root-cause?

**Comment by johnmelodyme** [comment]
> I had the same issue on a dotnet generate web project recently. But I can fix the issue by editing `Properties\launchSettings.json` to modify the `applicationUrl` in your app config.
> 
> From this
> 
> ```
> "applicationUrl": "https://localhost:5001;http://localhost:5000",
> ```
> to this
> 
> ```
> "applicationUrl": "http://localhost:5000",
> ```
> I suspect it could be an issue of the dotnet boilerplate project "ASP.NET Core Web API".

Still can't

**Comment by Sopcce** [comment]
on CentOS 7 ，I had the same issue on a dotnet generate web project recently.  
![image](https://user-images.githubusercontent.com/12589684/56738244-c9d23100-679e-11e9-9dc7-746e9503bcaf.png)
![image](https://user-images.githubusercontent.com/12589684/56738511-5bda3980-679f-11e9-94ff-e4ebc4ded35c.png)

```
{
  "iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:56271/",
      "sslPort": 0
    }
  },

  "profiles": {
    "IIS Express": {
      "commandName": "IISExpress",
      "launchBrowser": true,
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    },
    "Logging": {
      "LogLevel": {
        "Default": "Warning"
      }
    },
    "CategoryProjectCore": {
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "http://localhost:5000",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Production"
      }
    }
  },
  "ConnectionStrings": {
    "SopContext": "server=127.0.0.1;port=3306;database=sopcce;uid=root;pwd=123456;Charset=utf8mb4;SslMode = none;"
  }
}
```
I do not know ,try other way
[#948](https://github.com/dotnet/core/issues/948) not help me.

**Comment by karelz** [comment]
@Sopcce please do not comment on close issues. We rarely monitor them.
I would suggest to file a new issue with details - it would be good to summarize what EXACTLY you tried that did not help you. The issues are now too long and too messy to make heads and tails clearly. Thanks!

**Comment by jeffnyman** [comment]
If people are still commenting on a closed issue, it's probably because it's what they are most directed to which likely means the issue shouldn't be closed in the first place. We have 948 (closed), this ticket (closed). I'm not sure opening yet another one makes sense.

For those who want another experience report: this exact same issue currently happens to me all the time as well, with a fresh install of dotnet. The project is here:

https://github.com/jeffnyman/testing-aspnet-core

Running `dotnet run` works just fine with this project on Windows, not at all on Mac (either Sierra, High Sierra or Mojave).

It initially tells me port 5000 is in use. It's not; I've checked. But I did run the kill steps on anything on the port as provided by seba-i. I then added the following line to `Startup.cs`:

```
.UseUrls("http://*:5001")
```
When I do that, then it tells me port 5001 is already in use. I realize the issue is closed but a lot of people are going to probably end up here as they experiment.

**Comment by ricemeisterio** [comment]
> fix the issue by editing `Properties\launchSettings.json` to modify the `applicationUrl` in your app config.
> 
> From this
> 
> ```
> "applicationUrl": "https://localhost:5001;http://localhost:5000",
> ```
> 
> to this
> 
> ```
> "applicationUrl": "http://localhost:5000",
> ```
> 
> I suspect it could be an issue of the dotnet boilerplate project "ASP.NET Core Web API".

I just tried to create a new project on .net core 2.2 and had the same issue.
dotnet new mvc --auth Individual -o AspNetCoreTodo
dotnet run

making this change worked for me.

**Comment by osiem1984** [comment]
If you use Visual Studio Code you need to close application in terminal. Maybe you run your application using `dotnet run` command. Then it started. Even you close your tab in browser it still works. You need to go to terminal tab in your Visual Studio Code and then press "ctrl + c" to stop the application. After that you should have problem with starting it using debug or withoud debug command.

**Comment by tuliopaim** [comment]
I solved it here by deleting the "applicationUrl" in launchSettings.json
![image](https://user-images.githubusercontent.com/20960425/67628913-7a071c00-f84c-11e9-8409-c22394ef5b38.png)

**Comment by MayurSapaliga** [comment]
I created a new project in VSCode with dotnet new mvc command
Then run the app through terminal using dotnet run.
It just froze after a few info statements.
I ran using the debugger.
Gave me the error -  Failed to bind to address https://127.0.0.1:5001: address already in use.

Here's what I did - 
Stopped the debugger.
Hit Ctrl+C on terminal.
Reran using the Debugger.
Worked fine for me.

**Comment by IaamRich** [comment]
You need just Kill all networks.... adapters... and run again - then all will be good.

**Comment by tanifort** [comment]
I 'm not sure if someone else has been facing the same issue. If that is the case, simply end the .Net Core Task running on your computer. Run your dotnet App afterwards.

**Comment by Benja-blip** [comment]
None of the previous suggestions have worked for me so far. I've tried running and using ctrl + c in the terminal like MayurSapaliga suggested. I've edited the applicationUrl in the launchSettings.json file like tuliopaim suggested. I've edited the Program.cs file as suggested on the Microsoft docs page here https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/web-host?view=aspnetcore-3.1

None of this works. I tried changing the port my app runs on, but I keep getting "No webpage was found for the web address: http://localhost:5000/", and the same message for every other port I tried, which include 5002, 5003, 5004, 5005, 5050, 5051, and 5052. I have no idea which ports are the common second choice for such purposes. I basically just trying random ports I think aren't in use.

All of this happened because I started a new dotnet project. I ran it a couple of times, and then went back to the previous project I was working on, and now it absolutely will not run. I can go back to the other project (the new one I started), and it runs fine. It looks like VS Code just will not release the port. I've done ctrl + c both in VS Code and in the terminal. Interestingly enough, `dotnet run` works in VS Code (for the new program that's causing this whole mess), but the same command **won't** run in the terminal. I'll get the same error I get when trying to run my first program, 

Unable to start Kestrel.
System.IO.IOException: Failed to bind to address http://127.0.0.1:5000: address already in use.

I've shut down my computer and turned off my router as well. I've killed the processes running on the ports. Absolutely nothing works. What do you want from me, VS Code?!?!?!

Not sure how this would be relevant, but it may be worth mentioning that I'm running Linux Mint 18.

**Comment by vanessav02** [comment]
Eu troquei de porta, e funcionou

**Comment by dotnetryan** [comment]
For those running dotnet through the CLI (and linux) and not properly closing the process, I did basic CLI steps to kill dotnet. Type `ps aux | grep dotnet` or just `ps` and you'll get the PIDs of the kestrel processes running. Then type, `kill -9 <PID>` where `<PID>` is your process ID.

**Comment by codematrix** [comment]
Damn, I had my firewall turned off and when I turned it back on it started to work. Weird.

**Comment by mumby0168** [comment]
> I had the same issue on a dotnet generate web project recently. But I can fix the issue by editing `Properties\launchSettings.json` to modify the `applicationUrl` in your app config.
> 
> From this
> 
> ```
> "applicationUrl": "https://localhost:5001;http://localhost:5000",
> ```
> 
> to this
> 
> ```
> "applicationUrl": "http://localhost:5000",
> ```
> 
> I suspect it could be an issue of the dotnet boilerplate project "ASP.NET Core Web API".
This fixed my issue it looks like it tries to start this on http and https not sure though.

**Comment by AizenSousuke** [comment]
Kill dotnet host in Task Manager and it worked.

**Comment by utsabshrestha** [comment]
### So here is the solution, if you are working on linux
while running the **dotnet run** if you stop the server  using **ctr + z** then it _Ctrl+z_ sends the SIGTSTP (Signal Tty SToP) signal to the foreground job but don't actually stop the running kestrel server, so you have to mannualy end the process of the kestrel server.
**SOLLUTION**  Only use **ctr + c** to stop the kestrel server, actually its even displayed on the shell while you run dotnet run.

To mannually kill the kestrel server use this command to see the running process and it pid on the port 5000

sudo lsof -iTCP -sTCP:LISTEN -P | grep :5000

then kill it with

**Comment by magicdawn** [comment]
why make this simple thing so complicated ???
why not just provide a `server.listen(port)`  API, so not the framework don't need to care other stuffs. and in user land code, we can do the config stuff or hard code a port here

**Comment by rapuncia** [comment]
> I had the same issue on a dotnet generate web project recently. But I can fix the issue by editing `Properties\launchSettings.json` to modify the `applicationUrl` in your app config.
> 
> From this
> 
> ```
> "applicationUrl": "https://localhost:5001;http://localhost:5000",
> ```
> 
> to this
> 
> ```
> "applicationUrl": "http://localhost:5000",
> ```
> 
> I suspect it could be an issue of the dotnet boilerplate project "ASP.NET Core Web API".

that really helped! thank you

**Comment by kaopath** [comment]
> ### So here is the solution, if you are working on linux
> while running the **dotnet run** if you stop the server using **ctr + z** then it _Ctrl+z_ sends the SIGTSTP (Signal Tty SToP) signal to the foreground job but don't actually stop the running kestrel server, so you have to mannualy end the process of the kestrel server.
> **SOLLUTION** Only use **ctr + c** to stop the kestrel server, actually its even displayed on the shell while you run dotnet run.
> 
> To mannually kill the kestrel server use this command to see the running process and it pid on the port 5000
> 
> sudo lsof -iTCP -sTCP:LISTEN -P | grep :5000
> 
> then kill it with
> 
> sudo kill -9 PID

fix the issue. thanks

**Comment by jpayne0061** [comment]
My issue was caused by a **docker** container keeping the port in a `CLOSE_WAIT` state

Using Windows 10 with .NET Core 2.1 web api.

I tried killing the PID I found with `netstat -aon | more`, but received the message `ERROR: The process "xxxx" not found.`

After hours of investigation, `netstat -b` yielded a clue:
![image](https://user-images.githubusercontent.com/6244839/89113514-31fed080-d440-11ea-8274-55672e05941b.png)

Docker Desktop comes with a standalone Kubernetes server and client, so I tried restarting Docker Desktop.

Fixed it right away. Hope it helps someone !

**Comment by smallprogram** [comment]
This is a problem with the Windows operating system. You can use commands 
`netsh interface ipv4 show excludedportrange protocol=tcp`
to view the ports reserved by the system. Note that these reserved ports are not static. It will change with windows updates.You may find that there are 5000 or 5001 in the reserved ports

**Comment by TooEnsureDevelopment** [comment]
# Solved

So when using docker I said

```docker run -p 5000:5003 tag/name```

Now its listening on that port.

I typed

```docker ps

and guess what... the container was still up (for 17 hours lol)

```docker kill ContainerID```

helped then I could run it

**Comment by daisyberry** [comment]
thanks to you'll!. the problem is your https request is insecure all you need do is go to your web browser and click on the advance button then click continue, or check for a running project on your terminal stop it and the run it again  thank you

**Comment by Razrmax** [comment]
> I had the same issue on a dotnet generate web project recently. But I can fix the issue by editing `Properties\launchSettings.json` to modify the `applicationUrl` in your app config.
> 
> From this
> 
> ```
> "applicationUrl": "https://localhost:5001;http://localhost:5000",
> ```
> 
> to this
> 
> ```
> "applicationUrl": "http://localhost:5000",
> ```
> 
> I suspect it could be an issue of the dotnet boilerplate project "ASP.NET Core Web API".

That actually did help me, thanks!

**Comment by tyaff** [comment]
> "applicationUrl": "http://localhost:5000",

This worked! Thanks for the tip!

**Comment by integral2dev** [comment]
> http://localhost:5000

this worked for me. by changing the port

**Comment by Cooperdale** [comment]
This may happen when you're running Multiple Startup Projects.
You have to check that the single projects are not running on the same engine, such as IIS Express.

Let's say you have 2 projects running, if both are starting on IIS Express they'll both use port 5001 and the second to start will be blocked by the first.

To check, select every single project you need to run. After selecting each project as if it was a single startup, choose the engine (the light green arrow next to the project's name points to the engine).
There must not be two projects running both on IIS Express or you'll get the conflict.

After selecting a different engine for each project, just reconfigure the Multiple Startup as it was before and it should work.

**Comment by chillNZ** [comment]
This fixed my issue:

```
net stop winnat
docker start ...
net start winnat
```

See https://stackoverflow.com/questions/54010365/how-to-see-what-is-reserving-ephemeral-port-ranges-on-windows and https://github.com/microsoft/WSL/issues/5306.

**Comment by Seabizkit** [comment]
HI 

This works:
Select Kesteal as the project Run instead of IIS express from drop down.
      Website project (Start up)
            => Starts another Owin project hosting API

**This does not work:**
Select IIS express as the project Run instead of Kesteal from drop down.
      Website project (Start up)
            => Starts another Owin project hosting API

Throws and error saying it cant bind to 172.0.0.1:"random port"

Under both, tell it which ports to run under... IIS express is assigned for me,
 but it defo! not the port which the hosted Owin project is, plus it complains about some random port of which either are configure for.

Owin project hosting API has .UseKestrel();

Owin project hosting API: 
```
public static IHostBuilder CreateHostBuilder(string[] args)
        {
            return Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    var port = args[0].Replace("port=", "");
                    var sslport = args[1].Replace("sslport=", "");

                    webBuilder.UseStartup<Startup>()
                    .UseUrls($"http://localhost:{port}", $"https://localhost:{sslport}")
                    .UseKestrel();
                });
        }
```
This should work running under iis express but it does not.

**Comment by antman2** [comment]
The op's error message is very frustrating for people trying to debug ASP.NET web apps under Docker as well.

No matter which port number/s you choose to use in `ASPNETCORE_URLS` the same error message gets thrown in the debugger quoting the newly chosen port number/s. For example, using Visual Studio Code...

#### .vscode/launch.json
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Docker: Debug web app",
            "type": "docker",
            "request": "launch",
            "preLaunchTask": "docker-debug-task",
            "dockerServerReadyAction": {
                "uriFormat": "%s://localhost:%s/.well-known/openid-configuration"
            },
            "netCore": {
                "appProject": "${workspaceFolder}/projectName/projectName.csproj"
            }
        }
    ]
}
```

#### .vscode/tasks.json
```json
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "docker-build-task",
            "type": "docker-build",
            "dockerBuild": {
                "dockerfile": "${workspaceFolder}/projectName/Dockerfile",
                "context": "${workspaceFolder}/projectName",
                "tag": "project-name:dev",
                "buildArgs": {
                    "CONFIG": "Debug"
                }
            }
        },
        {
            "label": "docker-debug-task",
            "type": "docker-run",
            "dependsOn": "docker-build-task",
            "dockerRun": {
                "containerName": "project-name-dev",
                "env": {
                    "ASPNETCORE_URLS": "http://+:12345"
                },
                "image": "project-name:dev",
                "ports": [
                    {
                        "containerPort": 12345,
                        "hostPort": 44306
                    }
                ],
                "remove": true
            },
            "netCore": {
                "appProject": "${workspaceFolder}/projectName/projectName.csproj",
                "enableDebugging": true,
                "configureSsl": false
            }
        }
    ]
}
```

Since we're in the context of Docker containers the error message is complaining about port availability inside the containers, nothing to do with ports on the Docker host, and it's because the Visual Studio Remote Debugger is trying (and failing) to launch a second instance of the web application for debugging purposes. It cannot do this because a first instance of the web application has already been launched by the container image's `ENTRYPOINT`.

I was able to get VS Remote Debugging working correctly by preventing the first (natural) instance of the web application from launching from the `ENTRYPOINT`. This is achieved by setting `dockerRun.customOptions` to execute something else instead (since this is a `dotnet-alpine` image I'm using `ash` instead of `bash`):

#### .vscode/tasks.json
```json
            "dockerRun": {
                "customOptions": "--entrypoint /bin/ash",
            }
```

I wasted far too much time figuring this out. Hope it helps somebody else.

**Comment by arauujoiago** [comment]
I had the same issue, but only on docker environment... To resolve, i just needed add `"Urls": "http://localhost:5000",` in appsettings.json

**Comment by vinn2010** [comment]
@arauujoiago 's point worked for me on mac.

**Comment by badly-drawn-wizards** [comment]
Just putting the solution to my problem here: I had multiple appsettings (appsettings.json, appsettings.Development.json) with Kestrel endpoints, where one was meant to override the previous endpoint, but since I accidentally had different names, I had two different Kestrel endpoints trying to bind to the same port.

**Comment by Awire9966** [comment]
In my case Im on windows and nothing so far has worked

**Comment by ALIYU28033** [comment]
> This fixed my issue:
> 
> ```
> net stop winnat
> net start winnat
> ```
> 
> See https://stackoverflow.com/questions/54010365/how-to-see-what-is-reserving-ephemeral-port-ranges-on-windows and [microsoft/WSL#5306](https://github.com/microsoft/WSL/issues/5306).

This also solves the issue for me

**Comment by BahadIsmayilov** [comment]
I also had this problem, just enter properties and launchsettings and change my applicationUrl to "applicationUrl": "http://localhost:5000", so it has worked
